privacy-on-beam/pbeam/aggregations.go

```diff
@@ -450,6 +483,213 @@ func clampNegativePartitionsFloat64Fn(v beam.V, r float64) (beam.V, float64) {
+	}
+
+	if err != nil {
+		log.Exit(err)
+	}
+	return fn
+}
+
+func prepareAddPartitionsInt64Fn(partitionKey beam.X) (k beam.X, v int64) {
+	return partitionKey, 0
+}
+
+func prepareAddPartitionsFloat64Fn(partitionKey beam.X) (k beam.X, v float64) {
+	return partitionKey, 0
+}
+
+func prepareAddPartitionsMeanInt64Fn(partitionKey beam.X) (k beam.X, v []int64) {
+	return partitionKey, []int64{}
+}
+
+func prepareAddPartitionsMeanFloat64Fn(partitionKey beam.X) (k beam.X, v []float64) {
+	return partitionKey, []float64{}
+}
+
+
+type Partition struct {
+	PartitionKey string
+}
+
+// Function for sum and mean. Drop unspecified partitions.
+func correctPartitions(s beam.Scope, partitions []beam.PCollection, pcol PrivatePCollection, partitionT beam.EncodedType) beam.PCollection {
+	if len(partitions) == 1 {
+		partitionsCol := partitions[0]
+		partitionsMap := beam.Combine(s, newKVPartitionsHashMapFn(partitionT), partitionsCol)
+		return beam.ParDo(s, prunePartitionsFn, pcol.col, beam.SideInput{Input: partitionsMap})
+	}
+	return pcol.col
+}
+
+// Function for count and distinct_id. Drop unspecified partitions.
+func correctCountPartitions(s beam.Scope, partitions []beam.PCollection, pcol PrivatePCollection, partitionT typex.FullType) beam.PCollection {
+	if len(partitions) == 1 {
+		partitionsCol := partitions[0]
+		partitionsMap := beam.Combine(s, newPartitionsHashMapFn(partitionT), partitionsCol)
+		return beam.ParDo(s, newPreparePruneFn(partitionT), pcol.col, beam.SideInput{Input: partitionsMap})
+	}
+	return pcol.col
+}
+
+type mapAccum struct {
+	PartitionMap map[Partition]bool
+}
+
+// Makes hashmaps for partitions stored as values.
+type partitionsHashMapFn struct {
+	PartitionType         beam.EncodedType
+	partitionEnc          beam.ElementEncoder
+}
+
+func newPartitionsHashMapFn(partitionFullType typex.FullType) *partitionsHashMapFn {
+	return &partitionsHashMapFn{
+		PartitionType:         beam.EncodedType{partitionFullType.Type()},
+	}
+}
+
+func (fn *partitionsHashMapFn) Setup(){
+	fn.partitionEnc = beam.NewElementEncoder(fn.PartitionType.T)
+}
+
+func (fn *partitionsHashMapFn) CreateAccumulator() mapAccum {
+	return mapAccum{PartitionMap: make(map[Partition]bool)}
+}
+
+func (fn *partitionsHashMapFn) AddInput(m mapAccum, partitionKey beam.X) mapAccum {
+	var partitionBuf bytes.Buffer
+	if err := fn.partitionEnc.Encode(partitionKey, &partitionBuf); err != nil {
+		log.Exitf("pbeam.partitionsHashMapFn.ProcessElement: couldn't encode partition key %v: %v", partitionKey, err)
+	}
+	m.PartitionMap[Partition{PartitionKey: string(partitionBuf.Bytes())}] = true
+	return m
+}
+
+func (fn *partitionsHashMapFn) MergeAccumulators(a, b mapAccum) mapAccum {
+	for k := range a.PartitionMap {
+		b.PartitionMap[k] = true
+	}
+	return b
+}
+
+func (fn *partitionsHashMapFn) ExtractOutput(m mapAccum) map[Partition]bool {
+	return m.PartitionMap
+}
+
+// Makes hashmaps for partitions stored in kv pairs.
+type KVpartitionsHashMapFn struct {
```

Comment:
Why do we need both of these hash map functions? Can't we enforce that our partition list is always the same format?

Reply:
I don't think so, because the PCollection data is formatted differently... @miracvbasaran ?

Reply:
Just kidding, was able to just use one of them. Thanks John!